org.eclipse.lsp4xml.extensions.maven/src/main/java/org/eclipse/lsp4xml/extensions/maven/MavenCompletionParticipant.java

```diff
@@ -113,11 +121,46 @@ public void onXMLContent(ICompletionRequest request, ICompletionResponse respons
 					phase -> phase.description, request).forEach(response::addCompletionAttribute);
 			break;
 		case "groupId":
-			if (!parent.getParentElement().getLocalName().equals("parent") && !parent.getParentElement().getLocalName().equals("plugin")) {
-				collectSimpleCompletionItems(localRepositorySearcher.searchGroupIds(RepositorySystem.defaultUserLocalRepository),
+			if (isParentDeclaration) {
+				// TODO local
+				collectParentCompletion(request, response);
+			} else {
+				// TODO if artifactId is set and match existing content, suggest only matching groupId
+				collectSimpleCompletionItems(isPlugin ? localRepositorySearcher.searchPluginGroupIds() : localRepositorySearcher.searchGroupIds(),
 						Function.identity(), Function.identity(), request).forEach(response::addCompletionAttribute);
+				internalCollectRemoteGAVCompletion(request, isPlugin).forEach(response::addCompletionItem);
+			}
+			break;
+		case "artifactId":
+			if (isParentDeclaration) {
+				// TODO local
+				collectParentCompletion(request, response);
+			} else {
+				(isPlugin ? localRepositorySearcher.getLocalPluginArtifacts() : localRepositorySearcher.getLocalArtifactsLastVersion()).keySet().stream()
+					.filter(groupIdArtifactId -> groupId.isEmpty() || groupIdArtifactId.groupId.equals(groupId.get()))
+					.map(groupIdartifactId -> groupIdartifactId.artifactId)
+					.distinct()
+					// TODO pass description as documentation
+					// TODO complete groupId if missing
+					// TODO? complete version
+					.map(artifact -> toCompletionItem(artifact, null, request.getReplaceRange()))
+					.forEach(response::addCompletionItem);
+				internalCollectRemoteGAVCompletion(request, isPlugin).forEach(response::addCompletionItem);
 			}
 			break;
+		case "version":
+			if (!isParentDeclaration) {
+				if (!artifactId.isEmpty()) {
```

Comment:
Since artifactId is Optional<String> we can't directly access the String.
I suggest doing something like 	
`if ( artifactId.isPresent() && !artifactId.get().isEmpty()) {...`

Not entirely sure yet if the isEmpty() call becomes redundant.

Reply:
Since we build the optional with a "filter out if empty", we're sure that the value of the string we'll get is a non-empty one. Ie if the optional has content, we're sure the contained string isn't empty.

Reply:
Okay good point. However, the `.get()` call is still missing which is causing compilation errors I believe.